internal/utils/utils.go

```diff
@@ -0,0 +1,363 @@
+package utils
+
+import (
+	"context"
+	"fmt"
+	"github.com/go-logr/logr"
+	ecv1alpha1 "go.etcd.io/etcd-operator/api/v1alpha1"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/utils/ptr"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/apiutil"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"strconv"
+	"strings"
+	"time"
+)
+
+const (
+	SSSmallerThanDesired = "SSSmallerThanDesired"
+	SSBiggerThanDesired  = "SSBiggerThanDesired"
+	SSEqualToDesired     = "SSEqualToDesired"
+)
+
+func prepareOwnerReference(ec *ecv1alpha1.EtcdCluster, scheme *runtime.Scheme) ([]metav1.OwnerReference, error) {
+	gvk, err := apiutil.GVKForObject(ec, scheme)
+	if err != nil {
+		return []metav1.OwnerReference{}, err
+	}
+	ref := metav1.OwnerReference{
+		APIVersion:         gvk.GroupVersion().String(),
+		Kind:               gvk.Kind,
+		Name:               ec.GetName(),
+		UID:                ec.GetUID(),
+		BlockOwnerDeletion: ptr.To(true),
+		Controller:         ptr.To(true),
+	}
+
+	var owners []metav1.OwnerReference
+	owners = append(owners, ref)
+	return owners, nil
+}
+func CreateOrPatchSS(ctx context.Context, logger logr.Logger, ec *ecv1alpha1.EtcdCluster, c *client.Client, replicas int32, scheme *runtime.Scheme) error {
+	sts := &appsv1.StatefulSet{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      ec.Name,
+			Namespace: ec.Namespace,
+		},
+	}
+
+	labels := map[string]string{
+		"app":        ec.Name,
+		"controller": ec.Name,
+	}
+
+	if err := validateOwner(ec, sts); err != nil {
```

Comment:
You already set the StatefulSet's namespace = ec.Namespace above, I don't see any necessarity to check their namespaces match again. Also the name `validateOwner` isn't accurate.

So suggest to remove the function `validateOwner` to keep it as simple as possible for now.